src/include/buffer.h

```diff
@@ -1230,41 +1230,37 @@ inline namespace v15_2_0 {
     }
   };
 
-inline bool operator>(bufferlist& l, bufferlist& r) {
-  for (unsigned p = 0; ; p++) {
-    if (l.length() > p && r.length() == p) return true;
-    if (l.length() == p) return false;
-    if (l[p] > r[p]) return true;
-    if (l[p] < r[p]) return false;
-  }
+inline bool operator==(const bufferlist &lhs, const bufferlist &rhs) {
+  if (lhs.length() != rhs.length())
+    return false;
+  return std::equal(lhs.begin(), lhs.end(), rhs.begin(), rhs.end());
 }
-inline bool operator>=(bufferlist& l, bufferlist& r) {
-  for (unsigned p = 0; ; p++) {
-    if (l.length() > p && r.length() == p) return true;
-    if (r.length() == p && l.length() == p) return true;
-    if (l.length() == p && r.length() > p) return false;
-    if (l[p] > r[p]) return true;
-    if (l[p] < r[p]) return false;
+
+inline bool operator<(const bufferlist& lhs, const bufferlist& rhs) {
+  auto l = lhs.begin(), r = rhs.begin();
+  for (; l != lhs.end() && r != rhs.end(); ++l, ++r) {
+    if (*l < *r) return true;
```

Comment:
@cbodley two issues: if buffers are not the same sizes then it wouldn't work. Also, is this safe? Not sure what the `*` operator returns, but need to make sure that we don't accidentally read outside of the buffer.

Reply:
the bufferlist iterator has a value_type of char, so `operator*()` returns char and `operator++()` advances by one character. the iterators are safe to dereference because we've verified that they're `!= end()`